lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -872,87 +872,179 @@ class DecomposeAtenHardsigmoidOp : public OpRewritePattern<AtenHardsigmoidOp> {
-  Value prob =
-      rewriter.create<ConstantFloatOp>(loc, rewriter.getF64FloatAttr(p));
-  Value lb =
-      rewriter.create<ConstantFloatOp>(loc, rewriter.getF64FloatAttr(0.0));
-  Value ub =
-      rewriter.create<ConstantFloatOp>(loc, rewriter.getF64FloatAttr(1.0));
-
-  Value falseVal = rewriter.create<ConstantBoolOp>(loc, false);
-  Value noneVal = rewriter.create<ConstantNoneOp>(loc);
-
-  // Create a uniform random op with low and high set to lb and ub respectively.
-  Value uniformRandom = rewriter.create<PseudoAtenUniformOp>(
-      loc, inputType, input, lb, ub, noneVal);
-  Value gtValue =
-      rewriter.create<AtenLtScalarOp>(loc, boolType, uniformRandom, prob);
-  // Since `gtValue` will be a boolean tensor convert it back to the original
-  // type.
-  Value convertBack = rewriter.create<AtenToDtypeOp>(
-      loc, inputType, gtValue, convertIntVal, falseVal, falseVal, noneVal);
-  return convertBack;
+  // The `prob` is expected to be a float type tensor.
+  if (!probType.hasDtype() || !probType.getDtype().isa<mlir::FloatType>()) {
+    (void)rewriter.notifyMatchFailure(
+        op, "probabilities must be a float type tensor");
+    return nullptr;
+  }
+
+  // Since the `aten.rand_like` op expects float-type operand, create a
+  // float-type tensor with the same shape as that of the `input`.
+  Value none = rewriter.create<ConstantNoneOp>(loc);
+  Value constFalse = rewriter.create<ConstantBoolOp>(loc, false);
+  // `constDtypeF64` is an integer constant corresponding to the Float64 dtype.
+  int dtypeF64 = (int)getScalarTypeForType(rewriter.getF64Type());
+  Value constDtypeF64 =
+      rewriter.create<ConstantIntOp>(loc, rewriter.getI64IntegerAttr(dtypeF64));
+  auto floatResType = inputType.getWithSizesAndDtype(inputType.getSizes(),
+                                                     rewriter.getF64Type());
+  Value floatTensor = rewriter.create<AtenToDtypeOp>(
+      loc, floatResType, input, constDtypeF64, /*non_blocking=*/constFalse,
+      /*copy=*/constFalse, /*memory_format=*/none);
+  Value random = rewriter.create<AtenRandLikeOp>(
+      loc, floatResType, floatTensor, /*dtype=*/none, /*layout=*/none,
+      /*device=*/none, /*pin_memory=*/none, /*memory_format=*/none);
+
+  // Bernoulli(x, p) = rand_like(float(x)) < p.
+  auto boolResType = inputType.getWithSizesAndDtype(inputType.getSizes(),
+                                                    rewriter.getI1Type());
+  Value lessThanP =
+      rewriter.create<AtenLtTensorOp>(loc, boolResType, random, prob);
+
+  // `constDtypeInput` is an integer constant corresponding to `input` dtype.
+  int dtypeInput = (int)getScalarTypeForType(inputType.getDtype());
+  Value constDtypeInput = rewriter.create<ConstantIntOp>(
+      loc, rewriter.getI64IntegerAttr(dtypeInput));
+  // As the output is expected to be of the `input` type, convert the boolean
+  // tensor `lessThanP` to a `input` type tensor.
+  Value output = rewriter.create<AtenToDtypeOp>(
```

Comment:
I think we can create a new function for type conversion, `static Value convertDtype(Value, from ,to)`.